### ISSUE TITLE
AutoAimHorizontal

### ISSUE THREAD

**Issue by AmandaKline** [The-Charge/SirLancebot2016#1]
- [x] Mount NUC (Next Unit of Computing, the little computer)
- [x] Mount camara
- [x] Camara subsystem
  - [x] isTargetVisible();
  - [x] targetHorizontalOffset();
- [x] AutoAimHorizontal Command (just the basics)
  - [x] Needs a while-held button
- [x] Test

After you do all that....
- [x] What if target not seen? (See separate issue)

**Comment by AmandaKline** [comment]
Today's progress: computer was mounted, camara was mounted. The net tables can be seen from the driverstation.

**Comment by AmandaKline** [comment]
Additional To-Do:
- [x] Make attachment to mount camara
- [x] Mount camara
- [x] Make mounting mechanism for minicomputer

The calculations for adjustment should be made in robot code and not in vision code.

**Our chasis with n wheels is hard to turn**
- Right now needs a value of ~.3 to turn at all, and at that speed it takes off more than we would like it to
  - This difficulty will likely be augmented one the rest of the weight of our robot is added
- DON'T want to drive one forward and the other backward: this would double the unpredictability
- Try setting one side to some value and the other side to slightly more than said value in order to turn more precisely

## Code ideas:

**Comment by AmandaKline** [comment]
- [x] make button - whilePressed()

**Comment by Shalmezad** [comment]
How to approach CameraSubsystem/AutoAimHorizontal:
- [x] Constants
  - [x] Create a constants class
  - [x] Add a public static final double TOWER_ROTATION_FROM_STARTPOINT
  - [x] Make that constant = 0 for now
- [x] CameraSubsystem
  - [x] Add a method to see if the target is visible ( boolean isTargetVisible(); )
  - [x] Add a method to get the xRotation ( double getTargetXRotation(); )
- [ ] AutoAimHorizontal
  - [x] Constructor
    - [x] requires Robot.driveTrain (use robot builder for this!)
    - [x] Do _NOT_ require Robot.cameraSubsystem
  - [x] Initialize
    - [x] Call driveTrain's initPercentVBusMode() (there's no guarantee what mode you'll start in)
    - [x] Send a 0 to left/right side (stop the robot)
  - [x] Execute
    - [x] See if the target is visible
      - [x] If so, see what the xRotation is, and move the proper direction.
      - [ ] If not, pick one of the options below:
        - [ ] look at the gyro, and use a direction to get you to TOWER_ROTATION_FROM_STARTPOINT
        - [ ] Turn in a preset direction 
        - [x] Do nothing
  - [x] isFinished
    - [x] return false (This is a while held command, it'll be ended/interrupted by the button being release)
  - [x] end/interrupted
    - [x] Set both sides of the drivetrain to 0
- [x] OI
  - [x] Use RobotBuilder
  - [x] Set up a joystick button to call AutoAimHorizontal while held
- [x] Test
  - [x] Set up plybot with everything (NUC, camera, etc)
  - [x] Verify that the NUC's numbers are being read (CameraDebug _should_ still be the default command for the camera subsystem)
  - [x] Have someone hold a target
  - [x] See if plybot follows the target as the person goes around the plybot

**Comment by cole-s** [comment]
Completed the steps from the check list above, however, the code needs to be tested on the new plybot. The plybot does need modification before we can test the code.

**Comment by AmandaKline** [comment]
Merged to Staging

**Comment by AmandaKline** [comment]
Today it was tested and it seems to work. Most of the time. Now only occasionally the camara system will loose power or calculation capabilities while the wheels are driving. Also, the closed loop system needs tuning.

**Comment by AmandaKline** [comment]
- [x] Add deadband

**Comment by Travis-Team2619** [comment]
Not sure if this should belong here or in a new issue, but when we have the shooter on the robot, we need to get some values for a polynomial equation for the auto-shooter speed for the distance. @AmandaKline - Should a new issue be created?

**Comment by CrystalENVT** [comment]
The deadband yesterday seemed a bit touchy. We need to do a quick test with production bot to see if the deadband should be broadened, or if it was due to slight movements of my hands.

**Comment by CrystalENVT** [comment]
Seemed to be working as of yesterday. It was running properly for an image count of about 92000 (with the image count increasing by 20 every second). No crash was happening, and everything was running correctly. Just need to test with the Flux Capacitor on the robot for the NUC instead of my laptop strapped onto plybot
